dcicutils/misc_utils.py

```diff
@@ -368,112 +369,108 @@ def utc_today_str():
+    and admits more questions. So, for now at least, this is not a subclass of RateManager but a separate
+    implementation.
+    """
+
+    EARLIEST_TIMESTAMP = datetime.datetime(datetime.MINYEAR, 1, 1)  # maybe useful for testing
+
+    def __init__(self, *, lockout_seconds, safety_seconds=0, action="metered action", enabled=True, log=None):
+        """
+        Creates a LockoutManager that cooperates in assuring a guarded operation is only happens at a certain rate.
+
+        The rate is once person lockout_seconds. This is a special case of RateManager and might get phased out
+        as redundant, but has slightly different operations available for testing.
+
+        Args:
+
+        lockout_seconds int: A theoretical number of seconds allowed between calls to the guarded operation.
+        safety_seconds int: An amount added to interval_seconds to accommodate real world coordination fuzziness.
+        action str: A noun or noun phrase describing the action being guarded.
+        enabled bool: A boolean controlling whether this facility is enabled. If False, waiting is disabled.
+        log object: A logger object (supporting operations like .debug, .info, .warning, and .error).
+        """
+
+        # This makes it easy to turn off the feature
+        self.lockout_enabled = enabled
+        self.lockout_seconds = lockout_seconds
+        self.safety_seconds = safety_seconds
+        self.action = action
+        self._timestamp = self.EARLIEST_TIMESTAMP
+        self.log = log or logging
+
+    @property
+    def timestamp(self):
+        """The timestamp is read-only. Use update_timestamp() to set it."""
+        return self._timestamp
+
+    @property
+    def effective_lockout_seconds(self):
+        """
+        The effective time between calls
+
+        Returns: the sum of the lockout and the safety seconds
+        """
+        return self.lockout_seconds + self.safety_seconds
+
+    def wait_if_needed(self):
+        """
+        This function is intended to be called immediately prior to each guarded operation.
+
+        This function will wait (using time.sleep) only if necessary, and for the amount necessary,
+        to comply with rate-limiting declared in the creation of this LockoutManager.
+
+        NOTE WELL: It is presumed that all calls are coming from this source. This doesn't have ESP that would
+        detect or otherwise accommodate externally generated calls, so violations of rate-limiting can still
+        happen that way. This should be sufficient for sequential testing, and better than nothing for
+        production operation.  This is not a substitute for responding to server-initiated throttling protocols.
+        """
+        now = datetime.datetime.now()
+        # Note that this quantity is always positive because now is always bigger than the timestamp.
+        seconds_since_last_purge = (now - self._timestamp).total_seconds()
```

Comment:
This variable name looks specific to a particular use case.

Reply:
No, it's just a matter of making it debuggable. If time keeps changing out from underneath you while you're in the debugger, it's hard to step. It's not any different than doing `foo = computation()` where computation could change due to external factors and any attempt to put the result in a variable inevitably freezes the item for view at the risk that you're not picking up changes. But in return you get consistency.

Reply:
Oh, I see. The purge variable. RIght. Yeah, I'll fix that.

Reply:
Changing it to `seconds_since_last_attempt`.